### PR TITLE
Better icon guessing algo

### PR DESCRIPTION
**Pull request by Cod3dDOT**
Part of #258
Based on NOAA and pirate weather

Logic used to determine the icon:
1. If total precipitation > 1mm and pop > 30%: return the most common precipitation type (default to rain if unidentifiable)
	PS: If pop is not found, assume as 100%. 
	
2. If average visibility < 5km:
	If < 5km: return HAZE
	If < 1km: return FOG
		
3. If average winds > 10km/h: return WIND
	
4. Check for clouds (average cloud cover):
	If in 75-100: return CLOUDY
	If in 37.5-75: return PARTLY_CLOUDY 
	If in 0-37.5: return CLEAR

### CODE REVIEW

File: app/src/main/java/org/breezyweather/sources/CommonConverter.kt [hunk 1]
```diff
@@ -673,48 +678,190 @@ private fun completeHalfDayFromHourlyList(
...
-            .firstOrNull { it.date.time == dailyDate.time + (if (isDay) 12 else 24) * 3600 * 1000 }
-            ?: halfDayHourlyList.first() // Use first element in the list to avoid null
+        // best representative hour we strive to find
+        var hour: HourlyWrapper? = null
+        // custom weather code to override the day code if needed
+        var customWeatherCode: WeatherCode? = null
+        // ideal time (12:00 for day and 24:00 for night)
+        val bestTime = Date(dailyDate.time + (if (isDay) 12 else 24).hours.inWholeMilliseconds)
+        // sort by time elapsed from the ideal temporal hour
+        val sorted = halfDayHourlyList.sortedBy { abs(bestTime.time - it.date.time) }
```
Comment (papjul) [inline app/src/main/java/org/breezyweather/sources/CommonConverter.kt:719]: What is the purpose? All hours from half day are equally important

File: app/src/main/java/org/breezyweather/sources/CommonConverter.kt [hunk 2]
```diff
@@ -673,48 +678,190 @@ private fun completeHalfDayFromHourlyList(
...
+    val avgVisibility = halfDayHourlyList.mapNotNull { it.visibility }.let {
+        if (it.isNotEmpty()) it.average() else null
+    }
+
     var halfDayWeatherText = newHalfDay.weatherText
     var halfDayWeatherPhase = newHalfDay.weatherPhase
     var halfDayWeatherCode = newHalfDay.weatherCode
 
     // Weather code + Weather text
     if (halfDayWeatherCode == null || halfDayWeatherText == null) {
```
Comment (papjul) [inline app/src/main/java/org/breezyweather/sources/CommonConverter.kt:711]: Can you move icon computing to a separate function? Thanks!

File: app/src/main/java/org/breezyweather/sources/CommonConverter.kt [hunk 3]
```diff
@@ -673,48 +678,190 @@ private fun completeHalfDayFromHourlyList(
...
+        hour = hour ?: sorted.firstOrNull { it.weatherCode != null } ?: sorted.first()
 
+        // prioritize custom weather code if set
         if (halfDayWeatherCode == null) {
-            halfDayWeatherCode = halfDayHourlyListWeather.weatherCode
+            halfDayWeatherCode = customWeatherCode ?: hour.weatherCode
         }
         if (halfDayWeatherPhase == null) {
-            halfDayWeatherPhase = halfDayHourlyListWeather.weatherText
+            halfDayWeatherPhase = hour.weatherText
```
Comment (papjul) [inline app/src/main/java/org/breezyweather/sources/CommonConverter.kt:849]: We have R.array.live_wallpaper_weather_kinds that can provide text for each weather code, we shouldn't take the closest matching hour text as it applies only for that hour, and not for the whole half day. Example: weather text could be "Snow for 25 min", or things very specific to this hour.

Comment (papjul) [reply app/src/main/java/org/breezyweather/sources/CommonConverter.kt:849]: We should probably add a getWeatherText in WeatherViewController object.

**Review by papjul — CHANGES_REQUESTED** [review]

File: app/src/main/java/org/breezyweather/sources/CommonConverter.kt (hunk 1 above)
Comment (Cod3dDOT) [reply app/src/main/java/org/breezyweather/sources/CommonConverter.kt]: Removed

File: app/src/main/java/org/breezyweather/sources/CommonConverter.kt (hunk 2 above)
Comment (Cod3dDOT) [reply app/src/main/java/org/breezyweather/sources/CommonConverter.kt]: Done.

File: app/src/main/java/org/breezyweather/sources/CommonConverter.kt (hunk 3 above)
Comment (Cod3dDOT) [reply app/src/main/java/org/breezyweather/sources/CommonConverter.kt]: Unfortunately, did not find a beautiful solution for this. I have created `WeatherViewController.getWeatherText(WeatherCode)` which throws a NotImplementedException() for now.

Comment (Cod3dDOT) [reply app/src/main/java/org/breezyweather/sources/CommonConverter.kt]: If weatherText is missing, it is now autocompleted using R.array.live_wallpaper_weather_kinds (in WeatherViewController.getWeatherText)

**Comment by papjul** [comment]
Merged with 6bbac74c5f6c6cdbf722cb9755f70aa37ca6e07b

**Pull request closed without merging**